libbeat/publisher/pipeline/output_test.go

```diff
@@ -0,0 +1,260 @@
+				// Give some time for events to be published
+				timeout := 20 * time.Second
+
+				// Make sure that all events have eventually been published
+				return waitUntilTrue(timeout, func() bool {
+					return numEvents.Load() == client.Published()
+				})
+			}, nil)
+
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestPublishWithClose(t *testing.T) {
+	tests := map[string]func(uint) publishCountable{
+		"client":         newMockClient,
+		"network_client": newMockNetworkClient,
+	}
+
+	const minEventsInBatch = 50
+
+	for name, ctor := range tests {
+		t.Run(name, func(t *testing.T) {
+			seedPRNG(t)
+
+			err := quick.Check(func(i uint) bool {
+				numBatches := 1000 + (i % 100) // between 1000 and 1099
+
+				wqu := makeWorkQueue()
+				numEvents := atomic.MakeUint(0)
+
+				var wg sync.WaitGroup
+				for batchIdx := uint(0); batchIdx <= numBatches; batchIdx++ {
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+
+						batch := randomBatch(minEventsInBatch, 150, wqu)
+						numEvents.Add(uint(len(batch.Events())))
+						wqu <- batch
+					}()
+				}
+
+				// Publish at least 1 batch worth of events but no more than 20% events
+				publishLimit := uint(math.Min(minEventsInBatch, float64(numEvents.Load())*0.2))
+				client := ctor(publishLimit)
+				worker := makeClientWorker(nilObserver, wqu, client)
+
+				// Allow the worker to make *some* progress before we close it
+				timeout := 10 * time.Second
+				progress := waitUntilTrue(timeout, func() bool {
+					return client.Published() >= publishLimit
+				})
+				if !progress {
+					return false
+				}
+
+				// Close worker before all batches have had time to be published
+				err := worker.Close()
+				require.NoError(t, err)
+
+				published := client.Published()
+
+				// Start new worker to drain work queue
+				client = ctor(0)
+				makeClientWorker(nilObserver, wqu, client)
+				wg.Wait()
+
+				// Make sure that all events have eventually been published
+				timeout = 20 * time.Second
+				return waitUntilTrue(timeout, func() bool {
+					total := published + client.Published()
+					return numEvents.Load() == total
+				})
+			}, &quick.Config{MaxCount: 50})
+
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+type publishCountable interface {
+	outputs.Client
+	Published() uint
+}
+
+func newMockClient(publishLimit uint) publishCountable {
+	return &mockClient{publishLimit: publishLimit}
+}
+
+type mockClient struct {
+	mu           sync.RWMutex
+	publishLimit uint
+	published    uint
+}
+
+func (c *mockClient) Published() uint {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.published
+}
+
+func (c *mockClient) String() string { return "mock_client" }
+func (c *mockClient) Close() error   { return nil }
+func (c *mockClient) Publish(batch publisher.Batch) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Block publishing
+	if c.publishLimit > 0 && c.published >= c.publishLimit {
+		batch.Retry() // to simulate not acking
+		return nil
+	}
```

Comment:
@urso WDYT about this logic to emulate blocking publishing?

Reply:
This emulation is not blocking, but a failing output. The batch is not acked, but this Retrty + return `nil` will signal that the output is ready to consumer another batch.

 A Blocking simulation would require you to wait for some signal (e.g. via a control channel).

Reply:
At one point I had a sleep in here between the retry and return. The idea then was that the first client would be closed before the sleep finished.

A control channel is better than a sleep. Once the first client is closed I can close the control channel to remove the block. However, the Retry (before waiting to consume from the control channel) will still be needed, otherwise the final publish count doesn't add up to the expected total number of events.

Reply:
Implemented blocking with control channel in 508b60687.